### PR TITLE
Flip 'pull diagnostics' to be on by default.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
We've had this on internally for a while, and since all recent fixes have gone through, have not heard any complaints about the diagnostics experience.

We'd like to enable this early for everyone to get this out there for everyone to use.  Note: this comes with a feature-flag we can use to switch back to the non-pull (solutioncrawler model) in case we discover issues.  Once we've had enough time with people using the default (like a full point release), we can start removing the older systems.

PR: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/results?buildId=8713552&view=results
Build: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/512411

PR: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/results?buildId=8730199&view=results

### CODE REVIEW

**Comment by sharwell** [comment]
The text "experimental" needs to be removed from the option name at the same time.

**Comment by CyrusNajmabadi** [comment]
> The text "experimental" needs to be removed from the option name at the same time. 

Indeed!

**Comment by sharwell** [comment]
Filed #70042 based on initial impression when I tried turning this on

**Review by sharwell — DISMISSED** [review]
Requesting this be blocked on #65967. In its current form, LSP pull diagnostics does not work at all.

**Comment by sharwell** [comment]
I have been really trying to see a case where Pull Diagnostics works, but to date I haven't been able to use it for more than a few minutes before I reach a time box and have to turn it off to get work done. I'll give it another try this week, but I haven't seen any evidence to suggest this feature is ready for delivery.

**Comment by CyrusNajmabadi** [comment]
It's been on for the division for several months now, and we haven't heard about any issues so far. You've identified a couple of areas to improve parity, and we've taken them. We need to move forward with this, and can do so with all the internal and team testing we've gotten. 

There will probably still be things that arise, but we can address them as regular but fixes, versus anything that needs to hold up this feature.

**Comment by CyrusNajmabadi** [comment]
@dibarbet

**Review by dibarbet — COMMENTED** [review]
Should we wait to merge this until we have a 17.10 branch (since 17.9 is in stabilization)?  We can then immediately merge this in 17.10 P1 
https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/34944/Dev17.9-Schedule-Details

cc @arkalyanms for creating the 17.10 branch

**Comment by CyrusNajmabadi** [comment]
I'm ok with this going into 17.10

File: src/Features/LanguageServer/Protocol/Features/Options/DiagnosticOptionsStorage.cs [hunk 1]
```diff
@@ -4,17 +4,16 @@
...
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+internal sealed class DiagnosticOptionsStorage
 {
-    internal sealed class DiagnosticOptionsStorage
-    {
-        public static readonly Option2<bool> LspPullDiagnosticsFeatureFlag = new(
-            "dotnet_enable_lsp_pull_diagnostics", defaultValue: false);
+    public static readonly Option2<bool> PullDiagnosticsFeatureFlag = new(
+        "dotnet_enable_pull_diagnostics", defaultValue: true);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Features/Options/DiagnosticOptionsStorage.cs:12]: renaming this I think will reset it for people who previously changed the value.  Fine with that since we're switching the default, but FYI.

File: src/VisualStudio/Core/Def/Options/VisualStudioOptionStorage.cs [hunk 2]
```diff
@@ -240,7 +240,7 @@ public bool TryFetch(LocalUserRegistryOptionPersister persister, OptionKey2 opti
         {"dotnet_provide_date_and_time_completions", new RoamingProfileStorage("TextEditor.%LANGUAGE%.Specific.ProvideDateAndTimeCompletions")},
         {"dotnet_log_telemetry_for_background_analyzer_execution", new FeatureFlagStorage(@"Roslyn.LogTelemetryForBackgroundAnalyzerExecution")},
         {"dotnet_lightbulb_skip_executing_deprioritized_analyzers", new FeatureFlagStorage(@"Roslyn.LightbulbSkipExecutingDeprioritizedAnalyzers")},
-        {"dotnet_enable_lsp_pull_diagnostics", new FeatureFlagStorage(@"Lsp.PullDiagnostics")},
+        {"dotnet_enable_pull_diagnostics", new FeatureFlagStorage(@"Roslyn.EnablePullDiagnostics")},
```
Comment (dibarbet) [inline src/VisualStudio/Core/Def/Options/VisualStudioOptionStorage.cs:243]: `Lsp.PullDiagnostics` cannot be changed - it's the value we use in the targeted notification payload:
```
"FeatureFlags": {
    "Lsp.PullDiagnostics": 1
},
```

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorVerifierInProcess.cs [hunk 3]
```diff
@@ -130,6 +130,7 @@ public async Task CaretPositionAsync(int expectedCaretPosition, CancellationToke
                 [FeatureAttribute.Workspace, FeatureAttribute.SolutionCrawlerLegacy, FeatureAttribute.DiagnosticService, FeatureAttribute.ErrorSquiggles],
                 cancellationToken);
 
+            await Task.Delay(TimeSpan.FromSeconds(5));
```
Comment (dibarbet) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorVerifierInProcess.cs:133]: Seems suspicious - was this here because the editor doesn't provide an API for us to wait for diagnostics on their end?

If that is the case, we should likely have a 'WaitUntil function that waits until we have an expected value (with the normal cancellation token we use)

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/ErrorList_InProc.cs [hunk 4]
```diff
@@ -126,10 +126,17 @@ public static string GetDescription(this IVsTaskItem item)
...
-            ErrorHandler.ThrowOnFailure(hierarchy.GetProperty((uint)VSConstants.VSITEMID.Root, (int)__VSHPROPID.VSHPROPID_Name, out var name));
-            return (string)name;
+            try
+            {
+                var errorItem = (IVsErrorItem)item;
+                ErrorHandler.ThrowOnFailure(errorItem.GetHierarchy(out var hierarchy));
+                ErrorHandler.ThrowOnFailure(hierarchy.GetProperty((uint)VSConstants.VSITEMID.Root, (int)__VSHPROPID.VSHPROPID_Name, out var name));
+                return (string)name;
+            }
+            catch (Exception)
```
Comment (dibarbet) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/ErrorList_InProc.cs:136]: Why this change?

**Review by dibarbet — CHANGES_REQUESTED** [review]

File: src/Features/LanguageServer/Protocol/Features/Options/DiagnosticOptionsStorage.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Options/DiagnosticOptionsStorage.cs:12]: yup.  reverting this.

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorVerifierInProcess.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorVerifierInProcess.cs:133]: fixing up.

File: src/VisualStudio/Core/Def/Options/VisualStudioOptionStorage.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Options/VisualStudioOptionStorage.cs]: reverted.

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 5]
```diff
@@ -850,17 +850,17 @@ public static void Main()
...
             {
                 await testServices.ErrorList.ShowErrorListAsync(cancellationToken);
-                await Task.Delay(TimeSpan.FromSeconds(10));
+                string[] expectedContents = [$"Class1.cs(7, 13): {expectedSeverity} CS0168: The variable 'x' is declared but never used"];
 
-                string[] expectedContents =
-                [
-                    $"Class1.cs(7, 13): {expectedSeverity} CS0168: The variable 'x' is declared but never used",
-                ];
+                while (true)
```
Comment (dibarbet) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs:855]: I think there's a helper we can use - maybe https://sourceroslyn.io/#Microsoft.VisualStudio.IntegrationTest.Utilities/Helper.cs,95 (not sure if that still exists)

Comment (dibarbet) [reply src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs:855]: Also wonder if it would be good to either assert the cancellation token isn't None, or explicitly use the hang mitigating one

**Review by dibarbet — COMMENTED** [review]
Fine with this change assuming it merges into 17.10

**Review by dibarbet — APPROVED** [review]
speedometer passed, we still have the feature flag we can use to turn this off.
OK with this going in, just need to make sure the feature flag is sync'd with this default

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 6]
```diff
@@ -169,7 +163,7 @@ void RaiseSecond()
             await SetUpEditorAsync(markup, HangMitigatingCancellationToken);
             await TestServices.Editor.InvokeCodeActionListAsync(HangMitigatingCancellationToken);
             await TestServices.EditorVerifier.CodeActionAsync("Simplify delegate invocation", applyFix: true, ensureExpectedItemsAreOrdered: true, blockUntilComplete: true, cancellationToken: HangMitigatingCancellationToken);
-            await TestServices.Editor.PlaceCaretAsync("temp2", 0, 0, extendSelection: false, selectBlock: false, HangMitigatingCancellationToken);
+            await TestServices.Editor.PlaceCaretAsync("temp2(", 0, 0, extendSelection: false, selectBlock: false, HangMitigatingCancellationToken);
```
Comment (sharwell) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs:166]: Why did this change?

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 7]
```diff
@@ -824,15 +774,17 @@ await using (var fileChangeRestorer = await TestServices.Shell.PauseFileChangesA
...
-                string[] expectedContents =
-                [
-                    $"(Compiler) Class1.cs(7, 13): {expectedSeverity} CS0168: The variable 'x' is declared but never used",
-                ];
+                string[] expectedContents = [$"Class1.cs(7, 13): {expectedSeverity} CS0168: The variable 'x' is declared but never used"];
+
+                while (true)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    await Task.Delay(TimeSpan.FromSeconds(1));
```
Comment (sharwell) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs:782]: This is a problem. We need to wait directly for pull diagnostics to complete.

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/ErrorListInProcess.cs [hunk 8]
```diff
@@ -122,7 +121,7 @@ private static string GetMessage(ITableEntryHandle item)
                 var unknown => unknown.ToString(),
             };
 
-            var message = $"({source}) {document}({line + 1}, {column + 1}): {severity} {errorCode}: {text}";
```
Comment (sharwell) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/ErrorListInProcess.cs:125]: Why was this changed?

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs [hunk 9]
```diff
@@ -425,9 +425,19 @@ public async Task SetUseSuggestionModeAsync(bool forDebuggerTextView, bool value
...
+
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            var view = await GetActiveTextViewAsync(cancellationToken);
+            var viewTagAggregatorFactory = await GetComponentModelServiceAsync<IViewTagAggregatorFactoryService>(cancellationToken);
+
+            var aggregator = viewTagAggregatorFactory.CreateTagAggregator<IErrorTag>(view);
+            var tags = aggregator
+                .GetTags(new SnapshotSpan(view.TextSnapshot, 0, view.TextSnapshot.Length));
+
+            return tags.SelectAsArray(tag => new TagSpan<IErrorTag>(tag.Span.GetSpans(view.TextBuffer).Single(), tag.Tag));
```
Comment (sharwell) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs:440]: How is this different from what it did before?

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorVerifierInProcess.cs [hunk 10]
```diff
@@ -130,23 +130,50 @@ public async Task CaretPositionAsync(int expectedCaretPosition, CancellationToke
...
-
-                var actualTaggedText = actualTaggedSpan.Span.GetText();
-                Assert.Equal(expectedTag.taggedText, actualTaggedText);
-
-                AssertEx.NotNull(actualTaggedSpan.Tag.ToolTipContent);
-                var containerElement = (ContainerElement)actualTaggedSpan.Tag.ToolTipContent;
-                var actualTooltipText = CollectTextInRun(containerElement);
-                Assert.Equal(expectedTag.tooltipText, actualTooltipText);
+                cancellationToken.ThrowIfCancellationRequested();
+                await Task.Delay(TimeSpan.FromSeconds(1));
```
Comment (sharwell) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorVerifierInProcess.cs:136]: Need to wait directly for pull diagnostics

File: src/EditorFeatures/CSharpTest/InlineDiagnostics/InlineDiagnosticsTaggerProviderTests.cs [hunk 11]
```diff
@@ -52,6 +55,9 @@ private static async Task<ImmutableArray<ITagSpan<InlineDiagnosticsTag>>> GetTag
                 files: Array.Empty<string>(),
                 sourceGeneratedFiles: new[] { content },
                 composition: SquiggleUtilities.WpfCompositionWithSolutionCrawler);
+            Assert.True(workspace.TryApplyChanges(workspace.CurrentSolution.WithOptions(
+                workspace.CurrentSolution.Options.WithChangedOption(new OptionKey(DiagnosticOptionsStorage.PullDiagnosticsFeatureFlag), false))));
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InlineDiagnostics/InlineDiagnosticsTaggerProviderTests.cs:60]:
```suggestion
```

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs [hunk 12]
```diff
@@ -425,9 +425,19 @@ public async Task SetUseSuggestionModeAsync(bool forDebuggerTextView, bool value
             }
         }
 
-        public Task<ImmutableArray<TagSpan<IErrorTag>>> GetErrorTagsAsync(CancellationToken cancellationToken)
+        public async Task<ImmutableArray<TagSpan<IErrorTag>>> GetErrorTagsAsync(CancellationToken cancellationToken)
         {
-            return GetTagsAsync<IErrorTag>(cancellationToken);
+            // return GetTagsAsync<IErrorTag>(cancellationToken);
+
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs:431]:
```suggestion
```

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs:438]: the other codepath converts the tags to IMappingTagSpans, which these new editor tags do not impl.

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/ErrorListInProcess.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/ErrorListInProcess.cs:125]: 'source' is not a thing anymore.  we previously placed it in this info just so we could distinguish if we made the error item based on live items vs build items.

**Comment by CyrusNajmabadi** [comment]
@dibarbet this has been merged in.  We'll track the insertion to deal with targeted notifications.

**Comment by dibarbet** [comment]
submitted the targeted notification change to get CPS onto the same place
(internal only link - https://targetednotifications.vsdata.io/rule/e7b1535a-03d7-4d4e-8254-418517d57cfc)
